chain/client/proof.go

```diff
@@ -3,537 +3,526 @@ package rpc
+			return crypto.PubkeyToAddress(*pubuc).Bytes(), 27 + i, nil
+		}
+	}
+	return nil, 0, fmt.Errorf("No match address found")
+}
+
+func GetBlockHeaderMerkleParts(codec *codec.Codec, block *tmtypes.Header) BlockHeaderMerkleParts {
+	bhmp := BlockHeaderMerkleParts{}
+	bhmp.VersionAndChainIdHash = merkle.SimpleHashFromByteSlices([][]byte{
+		cdcEncode(codec, block.Version),
+		cdcEncode(codec, block.ChainID),
+	})
+	bhmp.TimeHash = merkle.SimpleHashFromByteSlices([][]byte{
+		cdcEncode(codec, block.Time),
+	})
+	bhmp.LastBlockIDAndOther = merkle.SimpleHashFromByteSlices([][]byte{
+		cdcEncode(codec, block.LastBlockID),
+		cdcEncode(codec, block.LastCommitHash),
+		cdcEncode(codec, block.DataHash),
+		cdcEncode(codec, block.ValidatorsHash),
+	})
+	bhmp.NextValidatorHashAndConsensusHash = merkle.SimpleHashFromByteSlices([][]byte{
+		cdcEncode(codec, block.NextValidatorsHash),
+		cdcEncode(codec, block.ConsensusHash),
+	})
+	bhmp.LastResultsHash = merkle.SimpleHashFromByteSlices([][]byte{
+		cdcEncode(codec, block.LastResultsHash),
+	})
+	bhmp.EvidenceAndProposerHash = merkle.SimpleHashFromByteSlices([][]byte{
+		cdcEncode(codec, block.EvidenceHash),
+		cdcEncode(codec, block.ProposerAddress),
+	})
+
+	return bhmp
+}
+
+func GetBlockRelayProof(cliCtx context.CLIContext, blockId uint64) (BlockRelayProof, error) {
+	bp := BlockRelayProof{}
+	_blockId := int64(blockId)
+	rc, err := cliCtx.Client.Commit(&_blockId)
+	if err != nil {
+		return BlockRelayProof{}, err
+	}
+	sh := rc.SignedHeader
+
+	block := *sh.Header
+	commit := *sh.Commit
+
+	bp.BlockHeaderMerkleParts = GetBlockHeaderMerkleParts(cliCtx.Codec, sh.Header)
+
+	leftMsg := ""
+	bp.Signatures = []TMSignature{}
+	addrs := []string{}
+	mapAddrs := map[string]TMSignature{}
+	for i, vote := range commit.Signatures {
+		msg := commit.VoteSignBytes(block.ChainID, i)
+		lr := strings.Split(hex.EncodeToString(msg), hex.EncodeToString(block.Hash()))
+
+		if leftMsg != "" && leftMsg != lr[0] {
+			return BlockRelayProof{}, fmt.Errorf("Inconsistent prefix signature bytes")
+		}
+		leftMsg = lr[0]
+
+		lr1, err := hex.DecodeString(lr[1])
+		if err != nil {
+			continue
+		}
+
+		addr, v, err := RecoverETHAddress(msg, vote.Signature, vote.ValidatorAddress)
+		if err != nil {
+			continue
+		}
+		addrs = append(addrs, string(addr))
+		mapAddrs[string(addr)] = TMSignature{
+			vote.Signature[:32],
+			vote.Signature[32:],
+			v,
+			lr1,
+		}
+	}
+	if len(addrs) == 0 {
+		return BlockRelayProof{}, fmt.Errorf("Too many invalid precommits")
+	}
+
+	sort.Strings(addrs)
+	for _, addr := range addrs {
+		bp.Signatures = append(bp.Signatures, mapAddrs[addr])
+	}
+
+	kb, err := hex.DecodeString(leftMsg)
+	if err != nil {
+		return BlockRelayProof{}, err
+	}
+	bp.SignedDataPrefix = kb
+
+	if err != nil {
+		return BlockRelayProof{}, err
+	}
+
+	return bp, nil
+}
+
+func MakeOtherStoresMerkleHash(mspo rootmulti.MultiStoreProofOp) (tmbytes.HexBytes, tmbytes.HexBytes, tmbytes.HexBytes) {
```

Comment:
Can we break this down into smaller, more understandable chunks that we can test separately?